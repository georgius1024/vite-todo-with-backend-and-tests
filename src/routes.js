import Todos from "./views/Todos.vue";
import Todo from "./views/Todo.vue";
import TodoNew from "./views/TodoNew.vue";

export default [
  { path: "/new", component: TodoNew },
  { path: "/:id", component: Todo },
  { path: "/", component: Todos },
];
