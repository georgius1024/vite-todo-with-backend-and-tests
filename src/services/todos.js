import axios from "axios";
const defaultUrl = 'https://jsonplaceholder.typicode.com'
export default function Todos(baseUrl = defaultUrl) {
  this.baseUrl = `${baseUrl}/todos`;
  this.url = function (id = null) {
    if (id) {
      return `${this.baseUrl}/${id}`;
    }
    return this.baseUrl;
  };
  this.request = function (url, method = "get", payload = null) {
    return axios[method](url, payload);
  };
  this.index = function () {
    return this.request(this.url());
  };
  this.show = function (id) {
    return this.request(this.url(id));
  };
  this.create = function (payload) {
    return this.request(this.url(), "post", payload);
  };
  this.update = function (id, payload) {
    return this.request(this.url(id), "put", payload);
  };
  this.destroy = function (id) {
    return this.request(this.url(id), "delete");
  };
}
