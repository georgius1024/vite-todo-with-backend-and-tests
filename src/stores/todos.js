import { defineStore } from "pinia";
import Todos from "../services/todos";
const RECORDS_PER_PAGE = 10;
export const useTodosStore = defineStore("todos", {
  state: () => ({ list: [], page: 0, service: new Todos() }),
  getters: {
    currentPage: (state) =>
      state.list.slice(
        state.page * RECORDS_PER_PAGE,
        (state.page + 1) * RECORDS_PER_PAGE
      ),
    maxPage: (state) =>
      Math.min(Math.ceil(state.list.length / RECORDS_PER_PAGE), 5),
  },
  actions: {
    async fetchAll() {
      const { data } = await this.service.index();
      this.list = data;
    },
    paginate(page) {
      this.page = page;
    },
    async update(todo) {
      this.list = this.list.map((e) => {
        if (e.id === todo.id) {
          return todo;
        }
        return e;
      });
      await this.service.update(todo.id, todo);
    },
    async create(todo) {
      this.list = [todo, ...this.list];
      await this.service.create(todo);
    },
    async destroy(todo) {
      this.list = this.list.filter((e) => (e.id !== todo.id));
      await this.service.destroy(todo.id);
    },
  },
});
