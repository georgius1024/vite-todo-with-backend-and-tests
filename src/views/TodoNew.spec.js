import { describe, it, expect, vi, beforeEach } from "vitest";
import { mount } from "@vue/test-utils";
import component from "./TodoNew.vue";
import { createTestingPinia } from '@pinia/testing'
import { createRouter, createWebHistory } from "vue-router";
import { flushPromises } from "@vue/test-utils";
import routes from "../routes";

const router = createRouter({
  history: createWebHistory(),
  routes,
});

vi.mock("../services/todos", () => ({
  default: function Todos() {
      this.index = vi.fn().mockResolvedValue({data: [{ id: 1 }, { id: 2 }, { id: 3 }]});
      this.show = vi.fn().mockResolvedValue(42);
      this.create = vi.fn().mockResolvedValue(42);
    },
}));
import { useTodosStore } from "../stores/todos";



describe("Todos page", () => {
  beforeEach(async () => {
    const todos = useTodosStore();
    await todos.fetchAll()
    router.push("/1");
    await router.isReady();
  });

  const global = {
    plugins: [router, createTestingPinia()]
  }
  it("mounts without error", async () => {
    const wrapper = mount(component, { global });
    await flushPromises()
    expect(wrapper.exists).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("save button emits", async () => {
    const wrapper = mount(component, { global });
    const push = vi.spyOn(router, 'push');
    await wrapper.find("input").setValue("hello");
    await wrapper.find("button.btn-primary").trigger("click");
    await flushPromises()
    expect(push).toBeCalledWith('/');
  });
});
