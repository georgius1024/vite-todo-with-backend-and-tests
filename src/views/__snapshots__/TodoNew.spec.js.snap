// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todos page > mounts without error 1`] = `
"<div>
  <h1>Create TODO Item</h1>
  <form><input type=\\"hidden\\">
    <div class=\\"mb-3\\"><label for=\\"exampleFormControlInput1\\" class=\\"form-label\\"> Enter title </label><input type=\\"text\\" class=\\"form-control\\"></div>
    <div class=\\"form-check my-3\\"><label class=\\"form-check-label\\"><input class=\\"form-check-input\\" type=\\"checkbox\\"> Completed </label></div>
    <div class=\\"mt-4\\"><button type=\\"button\\" class=\\"btn btn-outline-primary me-4\\"> Cancel </button><button type=\\"submit\\" class=\\"btn btn-primary me-4\\"> Save </button>
      <!--v-if-->
    </div>
  </form>
</div>"
`;
